### ISSUE TITLE
Tolerate USB debug tool theft - implement TargetController suspension

### ISSUE THREAD

**Issue by navnavnav** [navnavnav/Bloom#3]
The TargetController currently operates with the requirement of target & debug tool access and control. If, for whatever reason, the TC loses access to the debug tool or the connected target, an exception will likely be thrown during a subsequent IO operation, leading to the TC shutting down, and the rest of Bloom following.

The problem with this is that there are some programs that can "steal" a USB device right out of Bloom's hands. Even if Bloom has claimed an interface on it. These programs use an old version of libusb, which I suspect is the problem. I suspect a soft-reset is being issued, directly or indirectly, resulting in Bloom losing access to the device. What's worse is that there doesn't appear to be a way of being notified when this occurs. So the only point where we'll know this has happened, is when the TC attempts to perform another IO operation on the device, and catches a DeviceCommuncationFailure exception.

This may not seem like much of an issue, as surely there aren't many other programs that the user will want to use, with Bloom running, that will steal the debug tool from us, right? Well, yes, but I'm afraid avrdude is one of the known culprits. It will aggressively take ownership of the debug tool the moment the user runs it, causing the TC to shutdown and Bloom following suit.

Because of the issue described above, the TC needs to be able to tolerate the theft of the connected debug tool. The way I think this should work is: When the TC catches a DeviceCommuncationFailure exception, it should try to figure out if this was caused by a greedy program, or due to some other reason that's out of our control (e.g the user disconnecting the device). If it was some other issue, we can simply do as we currently do now - shutdown. But if it was caused by a greedy program, the TC should go into what I'm calling a "suspended state", where it release the device resources and sends the appropriate event to the DebugServer, to have it terminate any active debug session. In this suspended state, the TC should only listen for a few events. One of which should be the DebugSessionStarted event. When the DebugServer sends this event, we know that the user wants to begin another debug session, so at this point, the TC can leave the suspended state and attempt to reconnect to the debug tool. If successful, the TC should carry on as normal, but if it fails to reconnect to the device, then it should just shutdown.

Nothing is set in stone with the approach, but those are just my first thoughts on this. Another, similar approach would be to stop hogging the debug tool in the first place. So have the TC go into the suspended state the moment the active debug session ends. Something to think about.

**Comment by navnavnav** [comment]
Have made some good progress on this. Implemented & tested TargetController suspension - looking good.

I'm going to include a configuration parameter that will allow users to have the TC go into a suspended state, automatically, upon the wrap-up of a debug session. The TC will automatically resume upon a new debug session being kicked off.

I still need to implement correct handling of abrupt loss of hardware control due to other programs, but the foundation for this (TC suspension) is almost complete, so should be able to get on this soon.

**Comment by navnavnav** [comment]
Looking to have this ready for merging into the develop branch by the end of this bank holiday weekend.

Not sure when the release will be ready, as I have some other changes I want to make before issuing another release. And lots of testing to do.

**Comment by navnavnav** [comment]
Development for this is complete and the changes have been merged into the develop branch.

With these changes, users will be able to configure Bloom to release the debug tool at the end of the debug session. Bloom will automatically attempt to reconnect to the debug tool upon the start of a new debug session. This can be configured via the new `releasePostDebugSession` boolean parameter, which belongs in the new `debugTool` JSON object defined in each environment (see below for an example of the new config). The updated documentation for this will be published on the Bloom website with the new release.
```
{
  "environments": {
    "default": {
      "debugTool": {
        "name": "atmel-ice",
        "releasePostDebugSession": true
      },

      "target": {
        "name": "avr8",
        "physicalInterface": "debugWire"
      },

      "debugServer": {
        "name": "avr-gdb-rsp",
        "ipAddress": "127.0.0.1",
        "port": "1442"
      }
    }
  },

  "insight": {
    "enabled": true
  }
}
```

Included in these changes is the proper handling of debug tool theft and unexpected communication failures. With these changes, if Bloom encounters a communication failure with the debug tool, it will immediately put the TargetController into a suspended state and terminate any active debug session. As already mentioned, Bloom will automatically attempt to reconnect to the debug tool upon the start of a new debug session. If Bloom fails to reconnect to the debug tool at this point, it will terminate the debug session and shutdown.
This should cover cases where other programs steal the debug tool from Bloom - meaning users can freely run programs like avrdude to program their targets, without Bloom getting in the way or unexpectedly shutting down.

I'm not sure when these changes will be released. I've still got some other things I'd like to get into this release, including a refactor of part description files (now known as target description files or TDFs), TDF validation scripts and an investigation into a bug that may be an issue with avr-gdb.

**Comment by navnavnav** [comment]
These changes have been released in v0.2.0.

For more documentation on the new configuration for suspending the TC, see https://bloom.oscillate.io/docs/configuration#debug-tool-target-config
